src-cpp/server42/src/Workers.cpp

```diff
@@ -0,0 +1,324 @@
+                        } else {
+                            BOOST_LOG_TRIVIAL(error)
+                                << "drop packet by invalid payload size: " << udp->getLayerPayloadSize() << " "
+                                << "expected: " << sizeof(Payload);
+                        }
+                    } else {
+                        if (eth->getDestMac() != src_mac) {
+                            BOOST_LOG_TRIVIAL(error)
+                                << "drop packet by invalid mac dst: " << eth->getDestMac().toString() << " "
+                                << "expected :" << src_mac.toString();
+                        } else {
+                            BOOST_LOG_TRIVIAL(error)
+                                << "drop packet by missed udp layer";
+                        }
+                    }
+
+                    rte_pktmbuf_free(mbuf_table[i]);
+                }
+
+                if (bucket.packet_size()) {
+
+                    zmq::message_t message(bucket.ByteSizeLong());
+                    BOOST_LOG_TRIVIAL(debug) << "flow_bucket <" << bucket.DebugString() << ">";
+                    bucket.SerializeToArray(message.data(), message.size());
+
+                    /*
+                     * When a ZMQ_PUSH socket enters the mute state due to having reached the high water mark for all
+                     * downstream nodes, or if there are no downstream nodes at all, then any zmq_send(3) operations on
+                     * the socket shall block until the mute state ends or at least one downstream node becomes available
+                     * for sending; messages are not discarded.
+                     */
+                    while (!socket.send(message, ZMQ_DONTWAIT) && alive) {}
+
+                } else {
+                    BOOST_LOG_TRIVIAL(info) << "flow_bucket packet_size==0 " << bucket.DebugString() << "\n" << std::flush;
+                }
+
+            } catch (zmq::error_t &exception) {
+                BOOST_LOG_TRIVIAL(error) << "ZMQ Error " << exception.what();
+            } catch (std::exception &exception) {
+                BOOST_LOG_TRIVIAL(error) << "Error " << exception.what();
+            }
+        }
+
+        return true;
+    }
+
+
+    bool echo_thread(boost::atomic<bool> &alive,
+                     pcpp::DpdkDevice* device) {
+
+        pcpp::MBufRawPacket *rx_mbuf[Config::chunk_size] = {};
+
+        const pcpp::MacAddress src(device->getMacAddress());
+
+        std::map<std::string, uint64_t> fake_duration;
+        std::random_device rd{};
+        std::uniform_int_distribution<int> base_latency(10000, 20000);
+        std::uniform_int_distribution<int> random_latency(0, 1000);
+
+
+        while (alive.load()) {
+            uint16_t num_of_packets = device->receivePackets(rx_mbuf, Config::chunk_size, 0);
+
+            for (uint16_t i = 0; i < num_of_packets; ++i) {
+
+                pcpp::Packet parsed_packet(rx_mbuf[i]);
```

Comment:
how heavy this parse is? Since packets are predefined, does it make sence to work with offesets in buf instead of packet?

Reply:
not so have, but we have an optimized version of this code in docs.